Prefer `flat_map` to `inject`.

myronmarston opened the pull request:
It’s faster and simpler.

samphippen commented:
this is all LGTM.

yujinakayama commented on benchmarks/flat_map_vs_inject.rb:
```diff
@@ -0,0 +1,24 @@
+  end
+
+  x.report("inject") do
+    words.inject([]) { |a, w| a + w.codepoints }
```
For fair benchmark, I think this should be `a.concat(w.codepoints)` which doesn't create useless array instance for each iteration.

myronmarston replied:
That's a good point.  I added a case for `concat` in the next commit.  Looks like `flat_map` is still faster, but the overhead of our `FlatMap.flat_map` compatibility layer (necessary to support 1.8.7 w/o monkey patching `flat_map` on enumerable) makes that essentially even with `inject`/`concat`.

I think the flat map version reads nicer and will make it easy to transition to raw `flat_map` when we drop support for 1.8.7 (probably in RSpec 4), so I decided to stick with it.

cupakromer commented:
The implementation for [`RSpec::Core::FlatMap.flat_map`](https://github.com/rspec/rspec-core/blob/15de144e/lib/rspec/core/flat_map.rb) is incorrect for [1.8.7](https://github.com/rspec/rspec-core/blob/15de144e/lib/rspec/core/flat_map.rb#L11):

```ruby
array.map { |item| yield item }.flatten
```

MRI's [`flat_map`](http://ruby-doc.org/core-2.2.0/Enumerable.html#method-i-flat_map) isn't a full `flatten`. It is just a `concat` so we should use `flatten(1)`:

```ruby
array.map { |item| yield item }.flatten(1)
```

cupakromer commented:
I noticed that the `benchmark-ips` gem isn't in our `Gemfile`. Should it be?

myronmarston commented:
> The implementation for RSpec::Core::FlatMap.flat_map is incorrect for 1.8.7:

Good point. Corrected in cd52601a7fe0ea4c3f4f3f290f4336921afa11f7.

> I noticed that the benchmark-ips gem isn't in our Gemfile. Should it be?

I don't think it needs to be.  I have it in my `Gemfile-custom` (along with `relish`, `byebug` and `allocation_stats`).

Since `benchmark-ips` isn't needed by our build or for the normal things our contributors do, I don't think it needs to be in the `Gemfile`.  It's trivial for anyone who wants to write or run a benchmark to add it to their `Gemfile-custom`.

yujinakayama commented:
https://github.com/rspec/rspec-core/blob/15de144e/lib/rspec/core/flat_map.rb#L6-L8

```ruby
def flat_map(array)
  array.flat_map { |item| yield item }
end
```

AFAIK, in this case using `&block` should be faster.

```ruby
def flat_map(array, &block)
  array.flat_map(&block)
end
```

Ref: https://github.com/yujinakayama/astrolabe/blob/8b6d5babb42c95475ef07b727cac21fa2c74fb38/benchmark/performance_spec.rb#L65-L119

yujinakayama commented:
Yes, using `&block` removed most of the overhead.

```
Calculating -------------------------------------
            flat_map    11.481k i/100ms
          inject (+)     8.493k i/100ms
     inject (concat)    10.477k i/100ms
    FlatMap.flat_map    10.199k i/100ms
FlatMap.flat_map_via_block_pass
                        11.171k i/100ms
-------------------------------------------------
            flat_map    124.072k (± 5.4%) i/s -    619.974k
          inject (+)     91.350k (± 4.9%) i/s -    458.622k
     inject (concat)    116.069k (± 4.9%) i/s -    586.712k
    FlatMap.flat_map    117.650k (± 5.1%) i/s -    591.542k
FlatMap.flat_map_via_block_pass
                        125.773k (± 5.7%) i/s -    636.747k
```

myronmarston commented:
> AFAIK, in this case using &block should be faster.

That's odd.  I've always heard that `yield` is faster and our benchmarks confirm that:

https://github.com/rspec/rspec-core/blob/master/benchmarks/capture_block_vs_yield.rb

Any idea what causes the difference between your benchmark and ours?

yujinakayama commented:
> Any idea what causes the difference between your benchmark and ours?

`block.call` ([used in the benchmark](https://github.com/rspec/rspec-core/blob/4ecb96305355275cd5655e1da541ea0c35555c39/benchmarks/capture_block_vs_yield.rb#L11-L13) in rspec-core) and `&block` are different.

* Using `block.call` converts the given block to a `Proc` instance and then invokes `#call`.
* Defining `&block` as a method parameter and passing it to another method as `&block` directly passes the block without conversion. (I read about this somewhere, but cannot remember right now)

JonRowe commented:
This LGTM, can it be merged as is or do you want to factor in @yujinakayama's findings?

myronmarston commented:
> This LGTM, can it be merged as is or do you want to factor in @yujinakayama's findings?

I'm going to try to reproduce his findings in our own benchmarks first.

myronmarston commented:
OK, I updated our benchmarks.  The reality appears to be even more complicated than what @yujinakayama explained above:

* Capturing the block (e.g. `&block`) has a high constant cost, regardless of whether or not the block is used.  Using `&block` and `block.call` (or even `&block` and `yield`) is about 5x slower than a single simple `yield`.
* @yujinakayama is right that forwarding a captured block is much faster than using `block.call`.
* Whether or not `yield` is faster than forwarding a captured block appears to depend on the number of times the block is called.  My benchmark shows that the break-even point is at about 20-25 times....beyond that, and forwarding a captured block clearly wins, outweighing the constant cost of capturing `&block`.  By the time we get to calling the block 10K times, it's approaching twice as fast as using `yield`.
* For the specific case of `flat_map` it acts a bit different -- regardless of array size, I found that forwarding `&block` was faster than using `yield` in spite of the prior benchmark suggesting that would only happen with enumerables of 20+ items.  This really confuses me and I don't have a theory as to why this is.

I went ahead and changed to capturing-and-forwarding the block.

JonRowe commented:
LGTM so merged :)
